### PR TITLE
feat(github): permissions

### PR DESCRIPTION
**Pull request by olarclara**
Closes XTD-594.
Closes XTD-595.
Closes XTD-667.
Closes XTD-680.
Closes XTD-682.

## Import flow

- [x] As a user w/o any GitHub permissions.
- [x] As a user w/ permission for public repos only:
    - [x] See the error message when a repo import returns a 404.
    
n8fren-3000 preview csb app_dashboard_recent_workspace=411542e0-c5fb-4ee8-85c9-65c8cfd18e52

~Because we don't know if the repo was not found because it's private or if it doesn't exist at all, we can't provide a more specific error message. Currently, it says "Can't find a repository? Review your GitHub permissions". Should w change the copy?~

- [x] As a user w/ permission for public and private repos.
- [x] Check `/s` route.

## Preferences

- [x] As a user w/o any GitHub permissions.

n8fren-3000 preview csb app_dashboard (2)

~It currently says "Enables Committing & Pull Requests" but, as @tristandubbeld pointed out, it enables more things. I think "Repositories & Workflow" might fit the title better.~

---

- [x] As a user w/ permission for public repos only.
- [x] As a user w/ permission for public and private repos.

## Repositories

- [x] As a user w/o any GitHub permissions.
- [x] As a user w/ permission for public repos only.
- [x] As a user after dismissing the banner.

---

- [x] Check empty state:

n8fren-3000 preview csb app_dashboard (1)

~While this can be because all of that team's repos are private, @tristandubbeld raised that "Don't see all your repositories?" might be confusing since the page is empty.~

## Repository

- [x] As a user w/o any GitHub permissions viewing a repository (branches) page:

n8fren-3000 preview csb app_dashboard

~I've identified this scenario during implementation so there are no designs for it, the goal is to state that this repo (which is public) will be read-only unless the user grants permission to access their GitHub data. As @tristandubbeld raised, "Review your GitHub permissions to modify this repository. Review your GitHub permissions" is repetitive and we could benefit from better text.~

---

Sandboxes checks will be added in an upcoming PR.

### CODE REVIEW

**Comment by tristandubbeld** [comment]
Amazing job @olarclara! This makes granting access so much easier, and insights so much more clear. I went through all the scenarios, and have some findings:

**The create new modal**

> As a user w/o any GitHub permissions

I was able to grant access to public and private repositories at the same time, but the "access only public repositories" wasn't working for me the **first** time. This might have been a GitHub hiccup, because the second time it worked!

> As a user w/ permission for public repos only:
>   See the error message when a repo import returns a 404

I think the message was a bit confusing. If we know the user doesn't have private permission we could change it from "Can't find a repository?" to something like "Repository not found. You might not have the right permission to access the repository. Review your GitHub permissions".

**Repositories page**

> As a user w/ permission for public repos only

The empty state also says "Don't see all your repositories?" which might be a bit confusing for new users, as there are no repositories shown at all?

**Preferences**

> As a user w/o any GitHub permissions

The text accompanying the new changes says "Enables Committing & Pull Requests", but it enables more. Should we update this copy?

**Repository page**

> As a user w/o any GitHub permissions viewing a repository (branches) page

It says "Review your GitHub permissions to modify this repository. Review your GitHub permissions" which is a bit double to me. Should we change the first bit to something like "To make changes to this repository you'll need to grant access to CodeSandbox on GitHub"?

image

We might want to discuss these bits with @necoline and @lenasmayor too, and these are non-blocking for this PR, but I think we can make some improvements later!

File: packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+  if (restrictsPublicRepos) {
+    return (
+      <Stack direction="vertical" gap={6}>
+        <Stack direction="vertical" gap={0}>
+          <Text
+            css={{
+              marginBottom: '0.25rem',
+              color: 'rgba(255, 255, 255, 0.5)',
+              fontSize: '0.75rem',
```
[the remark this answers is not in the document]
Comment (tristandubbeld) [reply packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx:53]: Seeing more usages below in this component.

File: packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        {!restrictsPrivateRepos && <GitHubScope scope="private_repos" />}
+      </Stack>
+      {restrictsPrivateRepos ? (
+        <Button
+          onClick={() => signInGithubClicked('private_repos')}
+          variant="primary"
+          autoWidth
+        >
+          {' '}
+          Autorize access to private repositories{' '}
```
Comment (tristandubbeld) [inline packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx:156]: Are these `{' '}` necessary?

**Review by tristandubbeld — COMMENTED** [review]
A few comments (and nitpicks) here and there, but overall a lot of neat improvements! Let me know if you have questions about any of my comments :)

File: packages/app/src/app/components/CreateSandbox/Import/Import.tsx [hunk 3]
```diff
@@ -189,15 +198,22 @@ export const Import: React.FC<ImportProps> = ({ onRepoSelect }) => {
           <Button
             css={{ height: '32px', paddingRight: 24, paddingLeft: 24 }}
             aria-disabled={hasMaxPublicRepositories}
-            disabled={Boolean(url.error) || isLoading}
+            disabled={Boolean(url.error) || isLoading || disableImport}
             tabIndex={hasMaxPublicRepositories ? -1 : 0}
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/components/CreateSandbox/Import/Import.tsx]: Yes! Removed here: https://github.com/codesandbox/codesandbox-client/commit/7f5a7e2e2f21475f1a2f2b448781b22d834e0366

File: packages/app/src/app/components/CreateSandbox/Import/Import.tsx [hunk 4]
```diff
@@ -189,15 +198,22 @@ export const Import: React.FC<ImportProps> = ({ onRepoSelect }) => {
           <Button
             css={{ height: '32px', paddingRight: 24, paddingLeft: 24 }}
             aria-disabled={hasMaxPublicRepositories}
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/components/CreateSandbox/Import/Import.tsx]: Yes! Removed here: https://github.com/codesandbox/codesandbox-client/commit/7f5a7e2e2f21475f1a2f2b448781b22d834e0366

File: packages/app/src/app/components/CreateSandbox/Import/Import.tsx [hunk 5]
```diff
@@ -189,15 +198,22 @@ export const Import: React.FC<ImportProps> = ({ onRepoSelect }) => {
...
           </Button>
         </Stack>
-        <Element aria-atomic="true" id="form-error" role="alert">
+        <Element
+          aria-atomic="true"
+          css={{
+            marginTop: '8px',
+          }}
+          id="form-error"
+          role="alert"
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/components/CreateSandbox/Import/Import.tsx]: Adjusted here: https://github.com/codesandbox/codesandbox-client/pull/7407/commits/80b03a6db2d7dce552f9a26cbf73f7c4f368e198. Created a follow-up task here: https://linear.app/codesandbox/issue/CSB-5614/import-improve-form-feedback-for-screen-reader-users

File: packages/app/src/app/components/Integration/DetailInfo/DetailInfo.tsx [hunk 6]
```diff
@@ -19,11 +18,10 @@ export const DetailInfo: React.FC<IDetailInfoProps> = ({
   onSignIn,
 }) => (
   <Details>
-    <Margin right={2}>
+    <Stack direction="vertical" gap={0}>
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/components/Integration/DetailInfo/DetailInfo.tsx]: That's true! Fixed it here: 494ec63fc8ef3582be0e6204eaf9965061ef2d31.

File: packages/app/src/app/pages/Dashboard/Components/shared/githubPermissions.tsx [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            padding: `4px ${insideGrid ? 0 : '8px'}`,
+
+            '&:hover:not(:disabled), &:focus:not(:disabled)': {
+              color: '#E0E0E0',
+            },
+
+            '&:focus-within': {
+              outline: '#ac9cff solid 2px',
+              outlineOffset: '-2px',
+            },
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/pages/Dashboard/Components/shared/githubPermissions.tsx:98]: It should be focus-visible, thanks for catching it!

File: packages/app/src/app/pages/Dashboard/Components/shared/githubPermissions.tsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+          padding: `4px ${insideGrid ? 0 : '8px'}`,
+
+          '&:hover:not(:disabled), &:focus:not(:disabled)': {
+            color: '#E0E0E0',
+          },
+
+          '&:focus-within': {
+            outline: '#ac9cff solid 2px',
+            outlineOffset: '-2px',
+          },
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/pages/Dashboard/Components/shared/githubPermissions.tsx:169]: It should be focus-visible, thanks for catching it!

File: packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx (hunk 2 above)
Comment (olarclara) [reply packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx]: Nope, removed here: https://github.com/codesandbox/codesandbox-client/commit/746a1b1c010af07f7676d61348a65950a672744a.

File: packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/pages/common/Modals/PreferencesModal/Integrations/GitHubPermissions.tsx]: @tristandubbeld I had copied the elements from the legacy `Integration` component to keep the styles consistent. Replaced with pixels here: 53abf48099de814166af2f0e91985e971cc1e5c4.

File: packages/app/src/app/components/CreateSandbox/Import/useGithubRepo.ts [hunk 9]
```diff
@@ -45,9 +46,12 @@ export const useGithubRepo = ({
   }
 
   if (error) {
+    const _error = error.graphQLErrors[0];
+
     return {
       state: 'error',
-      error: error.message,
+      error: _error.message,
+      code: (_error as any)?.code,
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/components/CreateSandbox/Import/useGithubRepo.ts:54]: @tristandubbeld unfortunately, seems to be a mismatch between the error type the GraphQL client expects and the error type sent by the server. Didn't want to further investigate now to not increase the scope of this task.

File: packages/app/src/app/components/CreateSandbox/Import/Import.tsx [hunk 10]
```diff
@@ -7,11 +7,17 @@ import React from 'react';
...
 import { GithubRepoToImport } from './types';
 import { useGithubRepo } from './useGithubRepo';
 import { useImportAndRedirect } from './useImportAndRedirect';
 import { getOwnerAndRepoFromInput } from './utils';
-import { MaxPublicRepos, PrivateRepoFreeTeam } from './importLimits';
+import {
+  MaxPublicRepos,
+  PrivateRepoFreeTeam,
+  RestrictedPrivateReposImport,
+} from './importLimits';
```
[the remark this answers is not in the document]
Comment (olarclara) [reply packages/app/src/app/components/CreateSandbox/Import/Import.tsx]: @tristandubbeld you're right! I forgot that this goes against our dev guidelines. Adjusted here: 85a1de108657336c4eee389d30aca34051721a6e

**Comment by olarclara** [comment]
@tristandubbeld thank you for going through this PR! I've addressed the code comments and added screenshots + descriptions of the copy questions you have to the PR description.

> I was able to grant access to public and private repositories at the same time, but the "access only public repositories" wasn't working for me the first time. This might have been a GitHub hiccup because the second time it worked!

I wasn't able to repro this but I've seen GitHub take a while to respond, that could be it. I'll keep an eye out to see if it happens again.

**Comment by necoline** [comment]
Great comments @tristandubbeld. Below is my comments on the text suggestion

### The create new modal

**As a user w/ permission for public repos only:** 
✅ "Repository not found. You may be missing permissions needed to access the repository. Review your GitHub permissions".

### Repositories page

**As a user w/ permission for public repos only**: 
> The empty state also says "Don't see all your repositories?" which might be a bit confusing for new users, as there are no repositories shown at all?

I see Tristan's point here, maybe if we add 'all imported repos' it can be a cue to any user to consider if their permissions are blocking the ability to see repos that they _should_ be able to see.

✅ "Don't see all of your imported repositories?" 

### Preferences

**As a user w/o any GitHub permissions**

> The text accompanying the new changes says "Enables Committing & Pull Requests", but it enables more. Should we update this copy?

I agree that this text is confusing. It also slightly contradicts the text below it which talks about linking and exporting
Let's go with Tristan's suggestion for the title

✅ Repositories & Workflow

### Repository page

**As a user w/o any GitHub permissions viewing a repository (branches) page**

~"Review your GitHub permissions to modify this repository. Review your GitHub permissions"~

✅  "To make changes, you'll need to grant access to CodeSandbox. Review your GitHub permissions "?

**Comment by christianalfoni** [comment]
I can not see anything affecting Workspace Scopes directly 👍 😄 

Just a small comment on:

> I think the message was a bit confusing. If we know the user doesn't have private permission we could change it from "Can't find a repository?" to something like "Repository not found. You might not have the right permission to access the repository. Review your GitHub permissions".

When not having access we need to show a `404` or people can "test for repos". Meaning they can try urls they do not have access to and see if they exist. GitHub returns `404` if you do not have access, so we never really know if it exists or not and that is really how it should be 😄

**Review by tristandubbeld — APPROVED** [review]
Looks great, lets ship it 😎

**Pull request merged**